configuration/configurator/ConfigSchemaHelper.cpp

```diff
@@ -0,0 +1,898 @@
+    assert(m_pSchemaMapManager != NULL);
+
+    if (comp == NULL || *comp == 0)
+        return NULL;
+
+    CSchema* pSchema = NULL;
+
+    LOOP_THRU_BUILD_SET_MANAGER_BUILD_SET
+    {
+        const char *pSchemaName = CBuildSetManager::getInstance()->getBuildSetComponentFileName(idx);
+
+        if (pSchemaName != NULL && strcmp(comp, pSchemaName) == 0)
+        {
+             pSchema = m_pSchemaMapManager->getSchemaForXSD(pSchemaName);
+
+             assert(pSchema != NULL);
+
+             if (pSchema != NULL)
+             {
+                static StringBuffer strDoc;
+                strDoc.clear(); // needed when printing more than 1 component
+                pSchema->getDocumentation(strDoc);
+
+                return strDoc.str();
+             }
+        }
+    }
+    return NULL;
+}
+void CConfigSchemaHelper::printJSON(const char* comp, char **pOutput, int nIdx, bool bCleanUp) const
+{
+    if (! (comp != NULL && *comp != 0) )
+    {
+        DBGLOG("no component selected for JSON, index = %d", nIdx);
+        return;
+    }
+    assert(m_pSchemaMapManager != NULL);
+
+    StringBuffer strJSON;
+
+    strJSON.clear();
+    resetTables();
+
+    CSchema* pSchema = NULL;
+
+    LOOP_THRU_BUILD_SET_MANAGER_BUILD_SET
+    {
+        const char *pSchemaName = CBuildSetManager::getInstance()->getBuildSetComponentFileName(idx);
+        //const char *pSchemaName = CBuildSetManager::getInstance()->getBuildSetComponentName(idx);
+
+        if (pSchemaName != NULL && strcmp(comp, pSchemaName) == 0)
+        {
+            pSchema = m_pSchemaMapManager->getSchemaForXSD(pSchemaName);
+            assert(pSchema != NULL);
+
+            if (pSchema != NULL)
+            {
+                pSchema->getJSON(strJSON, 0, nIdx);
+                *pOutput = (char*)malloc((sizeof(char))* (strJSON.length())+1);
+
+                if (bCleanUp == true)
+                {
+                    this->clearLF(strJSON);
+                    strJSON.replaceString("\\","\\\\");
+                }
+
+                sprintf(*pOutput,"%s",strJSON.str());
+
+                return;
+            }
+            else
+            {
+                *pOutput = NULL;
+            }
+        }
+    }
+}
+
+void CConfigSchemaHelper::printJSONByKey(const char* key, char **pOutput, bool bCleanUp) const
+{
+    if (! (key != NULL && *key != 0) )
+    {
+        DBGLOG("no component key provided for to generate JSON");
+        return;
+    }
+
+    assert(m_pSchemaMapManager != NULL);
+
+    if (key[0] == '#')
+        key = &(key[1]);
+
+    StringBuffer strKey(key);
+    StringBuffer strJSON;
+
+    strJSON.clear();
+    resetTables();
+
+    const char *pChar = strrchr(key,'[');
+    assert(pChar != NULL && strlen(pChar) >= 3);
+
+    int length = strlen(pChar);
+
+    StringBuffer strIdx;
+
+    pChar++;
+
+    int nTemp = 0;
+
+    do
+    {
+        strIdx.append(*(pChar+nTemp));
+        nTemp++;
+    }while (*(pChar+nTemp) != ']');
```

Comment:
we might want to make sure sure pChar+nTemp != NULL as well. Aslo, we might not need the nTemp, and just increment pChar and test *pChar